python/ray/serve/storage/kv_store.py

```diff
@@ -0,0 +1,248 @@
+            raise TypeError("namespace must a string, got: {}.".format(
+                type(namespace)))
+
+        self.namespace = namespace or ""
+
+    def get_storage_key(self, key):
+        return "{ns}-{key}".format(ns=self.namespace, key=key)
+
+    def put(self, key, val):
+        """Put the key-value pair into the store.
+
+        Args:
+            key (str)
+            val (bytes)
+        """
+        if not isinstance(key, str):
+            raise TypeError("key must be a string, got: {}.".format(type(key)))
+        if not isinstance(val, bytes):
+            raise TypeError("val must be bytes, got: {}.".format(type(val)))
+
+        ray_kv._internal_kv_put(self.get_storage_key(key), val, overwrite=True)
+
+    def get(self, key):
+        """Get the value associated with the given key from the store.
+
+        Args:
+            key (str)
+
+        Returns:
+            The bytes value. If the key wasn't found, returns None.
+        """
+        if not isinstance(key, str):
+            raise TypeError("key must be a string, got: {}.".format(type(key)))
+
+        return ray_kv._internal_kv_get(self.get_storage_key(key))
+
+    def delete(self, key):
+        """Delete the value associated with the given key from the store.
+
+        Args:
+            key (str)
+        """
+
+        if not isinstance(key, str):
+            raise TypeError("key must be a string, got: {}.".format(type(key)))
+        return ray_kv._internal_kv_del(self.get_storage_key(key))
+
+
+class RayLocalKVStore(KVStoreBase):
+    """Persistent version of KVStoreBase for cluster fault
+    tolerance. Writes to local disk with provided path.
+
+    Supports only string type for key and bytes type for value,
+    caller must handle serialization.
+    """
+
+    def __init__(
+            self,
+            namepsace: str,
+    ):
+        self._namespace = namepsace
+        self._tombstone = b"\##DELETED\##"
+
+    def get_storage_key(self, key):
+        return "{ns}-{key}".format(ns=self._namespace, key=key)
+
+    def put(self, key, val):
+        """Put the key-value pair into the store.
+
+        Args:
+            key (str)
+            val (bytes)
+        """
+        if not isinstance(key, str):
+            raise TypeError("key must be a string, got: {}.".format(type(key)))
+        if not isinstance(val, bytes):
+            raise TypeError("val must be bytes, got: {}.".format(type(val)))
+
+        # For each {key, val} pair to put, we append a new line in each of
+        # ray_serve_checkpoint_key.txt and ray_serve_checkpoint_val.txt,
+        # with key string and btyes value appended respectively.
+        with open("/tmp/ray_serve_checkpoint_key.txt", "a+") as key_file, open(
+                "/tmp/ray_serve_checkpoint_val.txt", "ab+") as val_file:
+            key_file.write(self.get_storage_key(key) + "\n")
+            val_file.write(val)
+            val_file.write(b"\n")
```

Comment:
instead of inventing our own format, can we please just write it to sqlite? https://docs.python.org/3/library/sqlite3.html

Reply:
that's my first choice as well but i didn't find it anywhere else in ray oss import or testing ... to me it looks like we haven't made sqlite a dependency and ideally not adding it for simple case like this ?

Reply:
It's part of python stdlib

Reply:
You can see past example here https://github.com/ray-project/ray/pull/5937/files and find kv_store_service